src/factories/pid/PIDLookupTable_factory.h

```diff
@@ -0,0 +1,101 @@
+#include <edm4eic/ReconstructedParticleCollection.h>
+#include <edm4hep/utils/vector_utils.h>
+
+#include "services/pid_lut/PIDLookupTable_service.h"
+#include "extensions/jana/JOmniFactory.h"
+
+#include <random>
+
+namespace eicrecon {
+
+struct PIDLookupTableConfig {
+    std::string filename;
+    std::string url;
+};
+
+class PIDLookupTable_factory : public JOmniFactory<PIDLookupTable_factory, PIDLookupTableConfig> {
+
+private:
+    PodioInput<edm4hep::MCParticle> m_particles_in {this};
+    PodioOutput<edm4eic::ReconstructedParticle> m_recoparticles_out {this};
+
+    ParameterRef<std::string> m_filename {this, "filename", config().filename, "Relative to current working directory"};
+    ParameterRef<std::string> m_url {this, "url", config().url, "Only used if not found via filename"};
+    Service<PIDLookupTable_service> m_lut_svc {this};
+
+    std::mt19937 m_gen;
+    std::uniform_real_distribution<double> m_dist {0, 1};
+    const PIDLookupTable* m_lut;
+
+public:
+    void Configure() {
+        m_lut = m_lut_svc().FetchTable(m_filename(), m_url());
+    }
+
+    void ChangeRun(int64_t run_number) {
+    }
+
+    void Process(int64_t run_number, uint64_t event_number) {
+
+        // TODO: This is all very handwavy because I haven't been attending the PID datamodel discussions.
+        // Please look over this carefully and correct as needed!
+
+        for (const auto& mcparticle : *m_particles_in()) {
+
+            // Unpack lookup values from input
+
+            int pdg = mcparticle.getPDG();
+            int charge = mcparticle.getCharge();
+            double momentum = edm4hep::utils::magnitude(mcparticle.getMomentum());
+
+            // TODO: I'm still confused as to whether our lookup table actually contains eta vs theta.
+            double eta = edm4hep::utils::eta(mcparticle.getVertex());
+            double theta = edm4hep::utils::anglePolar(mcparticle.getVertex());
+            double phi = edm4hep::utils::angleAzimuthal(mcparticle.getVertex());
+
+            auto entry = m_lut->Lookup(pdg, charge, momentum, eta, phi);
+
+            int identified_pdg = -1; // unknown
+                                     // TODO: What is the PDG for 'unknown', actually?
+
+            if (entry != nullptr) {
+
+                m_gen.seed(event_number);
+                double random_unit_interval = m_dist(m_gen);
+
+                if (random_unit_interval < entry->prob_electron) {
+                    identified_pdg = 11; // electron
+                }
+                else if (random_unit_interval < (entry->prob_electron + entry->prob_pion)) {
+                    identified_pdg = 211; // pion
+                }
+                else if (random_unit_interval < (entry->prob_electron + entry->prob_pion + entry->prob_kaon)) {
+                    identified_pdg = 321; // kaon
+                }
+                else if (random_unit_interval < (entry->prob_electron + entry->prob_pion + entry->prob_kaon + entry->prob_electron)) {
```

Comment:
The approach of throwing one uniform value and comparing to the sum of the N probabilities is not exactly what we want and can do with the information provided. We should produce N random variables, which sum to 1, and where the expectation value of each is given by the entry in the table. Those N random variables are to represent the probability of each individual hypothesis.

The LUT doesn't have sufficient information to do this correctly, of course. We don't have the multivariate distributions for this random variable problem, only their summary statistic of the mean. But if this is intended to allow for a prototype that will get more precise, we should adhere to an interface that is what will be present in the future for full PID reconstruction.

Reply:
I don't agree. The statistical fluctuations of the original simulation define a systematic uncertatinty on the purities/backgrounds, the statistical fluctuations in the data define statistical uncertainty. Smearing, like you suggest, the fractions that we plan to use for the analysis is not a correct approach.